libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,551 @@
+				);
+		}
+		bool sourceIsSame(size_t _lhs, size_t _rhs)
+		{
+			return std::visit(util::GenericVisitor{
+				[&](PreviousSlot const&, PreviousSlot const&) { return true; },
+				[&](StackSlot const& _lhs, StackSlot const& _rhs) { return _lhs == _rhs; },
+				[&](auto const&, auto const&) { return false; }
+			}, layout.at(_lhs), layout.at(_rhs));
+		}
+		int sourceMultiplicity(size_t _offset)
+		{
+			return std::visit(util::GenericVisitor{
+				[&](PreviousSlot const&) { return 0; },
+				[&](StackSlot const& _s) { return multiplicity.at(_s); }
+			}, layout.at(_offset));
+		}
+		int targetMultiplicity(size_t _offset)
+		{
+			if (!outputs.count(post.at(_offset)) && !generateSlotOnTheFly(post.at(_offset)))
+				return 0;
+			return multiplicity.at(post.at(_offset));
+		}
+		bool targetIsArbitrary(size_t _offset)
+		{
+			return _offset < post.size() && std::holds_alternative<JunkSlot>(post.at(_offset));
+		}
+		void swap(size_t _i)
+		{
+			yulAssert(!holds_alternative<PreviousSlot>(layout.at(layout.size() - _i - 1)) || !holds_alternative<PreviousSlot>(layout.back()), "");
+			std::swap(layout.at(layout.size() -  _i - 1), layout.back());
+		}
+		size_t sourceSize() { return layout.size(); }
+		size_t targetSize() { return post.size(); }
+		void pop() { layout.pop_back(); }
+		void pushOrDupTarget(size_t _offset) { layout.push_back(post.at(_offset)); }
+	};
+
+	Shuffler<ShuffleOperations>::shuffle(_layout, _post, _generateSlotOnTheFly);
+
+	// Now we can construct the ideal layout before the operation.
+	// "layout" has the declared variables in the desired position and
+	// for any PreviousSlot{x}, x yields the ideal place of the slot before the declaration.
+	vector<optional<StackSlot>> idealLayout(_post.size(), nullopt);
+	for (auto const& [slot, idealPosition]: ranges::zip_view(_post, _layout))
+		if (PreviousSlot* previousSlot = std::get_if<PreviousSlot>(&idealPosition))
+			idealLayout.at(previousSlot->slot) = slot;
+
+	while (!idealLayout.empty() && !idealLayout.back())
+		idealLayout.pop_back();
+
+	return idealLayout | ranges::views::transform([](optional<StackSlot> s) {
+		yulAssert(s, "");
+		return *s;
+	}) | ranges::to<Stack>;
+}
+}
+
+Stack StackLayoutGenerator::propagateStackThroughOperation(Stack _exitStack, CFG::Operation const& _operation)
+{
+	Stack& stack = _exitStack;
+
+	// This is a huge tradeoff between code size, gas cost and stack size.
+	auto generateSlotOnTheFly = [&](StackSlot const&) {
+		//return stack.size() > 12 && canBeFreelyGenerated(_slot);
+		// return canBeFreelyGenerated(_slot);
+		return false;
+	};
+
+	size_t previousLayoutSize = stack.size();
+	for (auto const& slot: stack)
+		if (util::findOffset(_operation.output, slot) || generateSlotOnTheFly(slot))
+				--previousLayoutSize;
+
+	auto layout = ranges::views::iota(0u, previousLayoutSize) |
+		ranges::views::transform([](size_t _index) { return PreviousSlot{_index}; }) |
+		ranges::to<vector<variant<PreviousSlot, StackSlot>>>;
```

Comment:
```suggestion
	vector<variant<PreviousSlot, StackSlot>> layout;
	for (size_t i = 0; i < previousLayoutSize; ++i)
		layout.emplace_back(PreviousSlot{i});
```
although not sure which is better.

Reply:
Maybe combining to
```
vector<variant<PreviousSlot, StackSlot>> layout;
for (size_t i: ranges::views::iota(0u, previousLayoutSize))
    layout.emplace_back(PreviousSlot{i});
```
is best :-).